src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturation.cpp

```diff
@@ -35,68 +35,69 @@
  * @file ControlAllocationSequentialDesaturation.cpp
  *
  * @author Roman Bapst <bapstroman@gmail.com>
+ * @author Beat Küng <beat-kueng@gmx.net>
  */
 
 #include "ControlAllocationSequentialDesaturation.hpp"
 
 
-
 void
 ControlAllocationSequentialDesaturation::allocate()
 {
 	//Compute new gains if needed
 	updatePseudoInverse();
 
-	// Allocate
-	_actuator_sp = _actuator_trim + _mix * (_control_sp - _control_trim);
+	switch (_param_mc_airmode.get()) {
```

Comment:
It seems like you have completely removed the sequential desaturation code and packed a few methods instead.

- can you elaborate on what was wrong with the previous implementation?

- the intent of the module was to have one small class per allocation method, with one parameter to select the desired method. Here you have one class for 3 methods, and a total of two params. Can you refractor with one class per method?

Reply:
> can you elaborate on what was wrong with the previous implementation?

Sure:
- doing full mixing and unsaturating yaw first could lead to the following: a diagonal tilt request at low throttle where one actuator gets below 0 would lead to yaw unsaturating that actuator.
- I changed the priority order to unsaturate thrust last (essentially disabling airmode), but even then I managed to get non-zero outputs at zero throttle quite easily.

> the intent of the module was to have one small class per allocation method, with one parameter to select the desired method. Here you have one class for 3 methods, and a total of two params. Can you refractor with one class per method?

My reasoning here was:
- use the existing airmode param (easier transition & people know this already)
- maximize code sharing and as it's not a lot of code it's easier to understand having it all in one place

However I also don't mind much taking it apart.

Reply:
> doing full mixing and unsaturating yaw first could lead to the following: a diagonal tilt request at low throttle where one actuator gets below 0 would lead to yaw unsaturating that actuator.

Naive question, but is that a bad thing? If yaw has the lowest priority then it should be fine to change the yaw setpoint to unsaturate actuators right?

> I changed the priority order to unsaturate thrust last (essentially disabling airmode), but even then I managed to get non-zero outputs at zero throttle quite easily.

Yes I do not think this algorithm guaranties the absence of saturation. For this I think an interior method, like the one in my original PR (https://github.com/jlecoeur/Firmware/pull/1) is safer. I can try to bring it in.

Overall, this seems much more MC-specific than the sequential desaturation method. I think we should either keep the previous method and add the several airmodes as new options. Or kill sequential desaturation but be clear that the airmode stuff is not as generic. @RomanBapst you may have a word to say here as you wrote it.

Reply:
@bkueng @jlecoeur I had a look at this and I think it's fine what @bkueng did. The current code is still using the fundamental desaturaton technique which is what the contribution was about. When I implemented the sequential desaturation,  I did not pay much attention to the exact details of airmode so I just added a simple axis prioritization method. This was enough to demonstrate the functionality of the desaturation method but it's better if we reproduce the exact desired behavior of the old mixing logic which  was developed on the basis of real flight testing.

Reply:
Ok great

Reply:
> However I also don't mind much taking it apart.

@bkueng I think the multiple parameters are more problematic than the fact that the code resides in a single class. What about we keep both parameters for a while. If MC_AIRMODE (the default) is set to disabled, we use the CA_METHOD param (with a few values dedicated to airmodes), if AIRMODE is activated, we set CA_METHOD to the corresponding value. Then next release, we can remove MC_AIRMODE.